service/src/main/kotlin/fi/espoo/evaka/invoicing/service/InvoiceCorrectionQueries.kt

```diff
@@ -0,0 +1,133 @@
+// SPDX-FileCopyrightText: 2017-2024 City of Espoo
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+package fi.espoo.evaka.invoicing.service
+
+import fi.espoo.evaka.shared.ChildId
+import fi.espoo.evaka.shared.DaycareId
+import fi.espoo.evaka.shared.InvoiceCorrectionId
+import fi.espoo.evaka.shared.PersonId
+import fi.espoo.evaka.shared.db.Database
+import fi.espoo.evaka.shared.db.Predicate
+import fi.espoo.evaka.shared.db.PredicateSql
+import fi.espoo.evaka.shared.domain.FiniteDateRange
+import java.time.YearMonth
+import kotlin.math.abs
+
+private fun Database.Read.getInvoiceCorrections(where: Predicate): List<InvoiceCorrection> =
+    createQuery {
+            sql(
+                """
+SELECT
+    id,
+    target_month,
+    head_of_family_id,
+    child_id,
+    unit_id,
+    product,
+    period,
+    amount,
+    unit_price,
+    description,
+    note
+FROM invoice_correction
+WHERE ${predicate(where.forTable("invoice_correction"))}
+"""
+            )
+        }
+        .toList()
+
+fun Database.Read.getInvoiceCorrectionsForMonth(month: YearMonth): List<InvoiceCorrection> =
+    getInvoiceCorrections(Predicate { where("$it.target_month = ${bind(month)}") })
+
+fun Database.Read.getInvoiceCorrectionsByIds(
+    ids: Set<InvoiceCorrectionId>
+): List<InvoiceCorrection> =
+    getInvoiceCorrections(Predicate { where("$it.id = ANY(${bind(ids)})") })
+
+fun Database.Read.getInvoiceCorrectionsForHeadOfFamily(
+    personId: PersonId
+): List<InvoiceCorrection> =
+    getInvoiceCorrections(where = Predicate { where("$it.head_of_family_id = ${bind(personId)}") })
+        .sortedWith(
+            compareByDescending<InvoiceCorrection> { it.targetMonth }
+                .thenByDescending { abs(it.amount * it.unitPrice) }
+        )
+
+fun Database.Transaction.movePastUnappliedInvoiceCorrections(
+    headOfFamilyIds: Set<PersonId>?,
+    targetMonth: YearMonth,
+) {
+    val firstOfMonth = targetMonth.atDay(1)
+    val headOfFamilyPredicate =
+        if (headOfFamilyIds != null) {
+            PredicateSql { where("head_of_family_id = ANY (${bind(headOfFamilyIds)})") }
+        } else {
+            PredicateSql.alwaysTrue()
+        }
+
+    execute {
+        sql(
+            """
+UPDATE invoice_correction
+SET target_month = ${bind(firstOfMonth)}
```

Comment:
could also bind YearMonth directly since we have jdbi mapper for it?

Reply:
Fixed 👍